README.md
# docker-nginx-web2py-min-mysql

[![Build Status](https://travis-ci.com/onezoom/docker-mysql.svg?branch=master)](https://travis-ci.com/onezoom/docker-mysql)
[![Layers](https://images.microbadger.com/badges/image/onezoom/docker-mysql.svg)](http://microbadger.com/images/onezoom/docker-mysql)

Docker container for nginx+uwsgi with web2py using python 3.8 on Ubuntu 20.04 and a MySQL server, based on [onezoom/docker-nginx-web2py-min](https://github.com/onezoom/docker-nginx-web2py) with MySQL Server layer based on [madharjan/docker-mysql](https://github.com/madharjan/docker-mysql/)

## Features

* Environment variables to set web2py admin password
* User-provided appconfig.ini file can be specified
* Environment variables to create database, user and set password
* Bats [bats-core/bats-core](https://github.com/bats-core/bats-core) based test cases

## MySQL Server 5.7 (docker-nginx-web2py-min-mysql)

### Environment

| Variable                  | Default | Example                                                                |
|---------------------------|---------|------------------------------------------------------------------------|
| DISABLE_MYSQL             | 0       | 1 (to disable)                                                         |
| MYSQL_DATABASE            | temp    | mydb                                                                   |
| MYSQL_USERNAME            | mysql   | myuser                                                                 |
| MYSQL_PASSWORD            | mysql   | mypass                                                                 |
|                           |         |                                                                        |
| WEB2PY_ADMIN              |         | Pa55w0rd                                                               |
| DISABLE_UWSGI             | 0       | 1 (to disable)                                                         |
|                           |         |                                                                        |
| INSTALL_PROJECT           | 0       | 1 (to enable)                                                          |
| PROJECT_GIT_REPO          |         | [https://github.com/OneZoom/OZtree](https://github.com/OneZoom/OZtree) |
| PROJECT_GIT_TAG           | HEAD    | v5.1.4                                                                 |
| PROJECT_APPCONFIG_INI_PATH|         | /etc/appconfig.ini                                                     |


## Build

```bash
# clone project
git clone https://github.com/onezoom/docker-nginx-web2py-min-mysql
cd docker-mysql

# build
make

# tests
make run
make test

# clean
make clean
```

## Run

**Note**: update environment variables below as necessary

```bash
# prepare foldor on host for container volumes
sudo mkdir -p /opt/docker/mysql/etc/conf.d
sudo mkdir -p /opt/docker/mysql/lib/
sudo mkdir -p /opt/docker/mysql/log/

# stop & remove previous instances
docker stop mysql
docker rm mysql

# run container
docker run -d \
  -e MYSQL_DATABASE=mydb \
  -e MYSQL_USERNAME=user \
  -e MYSQL_PASSWORD=pass \
  -p 3306:3306 \
  -v /opt/docker/mysql/etc/conf.d:/etc/mysql/conf.d \
  -v /opt/docker/mysql/lib:/var/lib/mysql \
  -v /opt/docker/mysql/log:/var/log/mysql \
  --name mysql \
  onezoom/docker-nginx-web2py-min-mysql:5.7
```

## Systemd Unit File

**Note**: update environment variables below as necessary

```txt
[Unit]
Description=MySQL Server

After=docker.service

[Service]
TimeoutStartSec=0

ExecStartPre=-/bin/mkdir -p /opt/docker/mysql/etc/conf.d
ExecStartPre=-/bin/mkdir -p /opt/docker/mysql/lib
ExecStartPre=-/bin/mkdir -p /opt/docker/mysql/log
ExecStartPre=-/usr/bin/docker stop mysql
ExecStartPre=-/usr/bin/docker rm mysql
ExecStartPre=-/usr/bin/docker pull onezoom/docker-nginx-web2py-min-mysql:5.7

ExecStart=/usr/bin/docker run \
  -e MYSQL_DATABASE=mydb \
  -e MYSQL_USERNAME=user \
  -e MYSQL_PASSWORD=pass \
  -p 3306:3306 \
  -v /opt/docker/mysql/etc/conf.d:/etc/mysql/conf.d \
  -v /opt/docker/mysql/lib/:/var/lib/mysql \
  -v /opt/docker/mysql/log:/var/log/mysql \
  --name mysql \
  onezoom/docker-nginx-web2py-min-mysql:5.7

ExecStop=/usr/bin/docker stop -t 2 mysql

[Install]
WantedBy=multi-user.target
```

## Generate Systemd Unit File

| Variable            | Default          | Example                                                          |
|---------------------|------------------|------------------------------------------------------------------|
| PORT                | 3306             | 8080                                                             |
| VOLUME_HOME         | /opt/docker      | /opt/data                                                        |
| NAME                | mysql            | docker-mysql                                                           |
| MYSQL_DATABASE      | temp             | mydb                                                             |
| MYSQL_USERNAME      | mysql            | user                                                             |
| MYSQL_PASSWORD      | mysql            | pass                                                             |

```bash
# generate mysql.service
docker run --rm \
  -e PORT=3306 \
  -e VOLUME_HOME=/opt/docker \
  -e NAME=docker-mysql \
  -e MYSQL_DATABASE=mydb \
  -e MYSQL_USERNAME=user \
  -e MYSQL_PASSWORD=pass \
  onezoom/docker-nginx-web2py-min-mysql:5.7 \
  mysql-systemd-unit | \
  sudo tee /etc/systemd/system/mysql.service

sudo systemctl enable mysql
sudo systemctl start mysql
```


Changelog.md
## 5.7

### Changes

* Rebased on onezoom/docker-nginx-web2py-min:20.04_2.21.1
* Added pymysql python library
* Based on madharjan/docker-base:16.04
* Using MySQL Server 5.7.26
* Added `mysql-systemd-unit` script to generate systemd unit files

## 5.5

### Changes

* Based on madharjan/docker-base:14.04
* Using MySQL Server 5.5


test/test.sh
#!/bin/bash
set -e

PWD=`pwd`

echo " --> Starting container"
ID=`docker run -d -v $PWD/test:/test $NAME:$VERSION`
sleep 10

echo " --> Checking whether all services are running..."
docker exec -t $ID /bin/bash /test/status.sh
sleep 1

echo " --> Stopping container"
docker stop $ID >/dev/null
docker rm $ID >/dev/null


services/mysql-startup.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

MYSQL_DATABASE=${MYSQL_DATABASE:-temp}
MYSQL_USERNAME=${MYSQL_DB_USERNAME:-mysql}
MYSQL_PASSWORD=${MYSQL_DB_PASSWORD:-mysql}

MYSQL_CHARSET=${MYSQL_CHARSET:-utf8}
MYSQL_COLLATION=${MYSQL_COLLATION:-utf8_unicode_ci}

MYSQL_USER=mysql
MYSQL_DATA_DIR=${MYSQL_DATA_DIR:-/var/lib/mysql}

mkdir -p ${MYSQL_DATA_DIR}
chmod -R 0700 ${MYSQL_DATA_DIR}
chown -R ${MYSQL_USER}:${MYSQL_USER} ${MYSQL_DATA_DIR}

DISABLE_MYSQL=${DISABLE_MYSQL:-0}

if [ ! "${DISABLE_MYSQL}" -eq 0 ]; then
  touch /etc/service/mysql/down
else
  rm -f /etc/service/mysql/down
fi

if [ ! -f "/etc/mysql/conf.d/mysqld-skip-name-resolv.cnf" ]; then
  cp /config/etc/mysql/conf.d/mysqld-skip-name-resolv.cnf /etc/mysql/conf.d/mysqld-skip-name-resolv.cnf
fi

if [ ! -f "/etc/mysql/conf.d/mysqld-bind-address.cnf" ]; then
  cp /config/etc/mysql/conf.d/mysqld-bind-address.cnf /etc/mysql/conf.d/mysqld-bind-address.cnf
fi

# Blank password for debian-sys-maint user
sed 's/password = .*/password =  /g' -i /etc/mysql/debian.cnf

if [ ! -d ${MYSQL_DATA_DIR}/mysql ]; then
  echo "Initializing database ..."
  mysql_install_db --user=${MYSQL_USER} >/dev/null 2>&1
  echo "Starting MySQL Server ..."
  /usr/bin/mysqld_safe >/dev/null 2>&1 &

  timeout=30
  echo -n "Waiting for database server to accept connections"
  while ! /usr/bin/mysqladmin -u root status >/dev/null 2>&1
  do
    timeout=$(($timout - 1))
    if [ $timeout -eq 0 ]; then
      echo -e "\nCould not connect to database server. Aborting..."
      exit 1
    fi
  done
  echo

  echo "Creating debian-sys-maint user ..."
  mysql -uroot -e "GRANT ALL PRIVILEGES on *.* TO 'debian-sys-maint'@'localhost' IDENTIFIED BY '' WITH GRANT OPTION;"

  /usr/bin/mysqladmin --defaults-file=/etc/mysql/debian.cnf shutdown
fi

if [ -n "${MYSQL_DATABASE}" -o -n "${MYSQL_USERNAME}" ]; then
  echo "Starting MySQL Server ..."
  /usr/bin/mysqld_safe >/dev/null 2>&1 &

  timeout=30
  echo -n "Waiting for database server to accept connections"
  while ! /usr/bin/mysqladmin -u root status >/dev/null 2>&1
  do
    timeout=$(($timout - 1))
    if [ $timeout -eq 0 ]; then
      echo -e "\nCould not connect to database server. Aborting..."
      exit 1
    fi
  done
  echo

  if [ -n "${MYSQL_DATABASE}" ]; then
    echo "Creating database ..."
    mysql --defaults-file=/etc/mysql/debian.cnf \
      -e "CREATE DATABASE IF NOT EXISTS \`${MYSQL_DATABASE}\` DEFAULT CHARACTER SET \`${MYSQL_CHARSET}\` COLLATE \`${MYSQL_COLLATION}\`;"

    if [ -n "${MYSQL_USERNAME}" ]; then
      echo "Granting access to database \"${MYSQL_DATABASE}\" for user \"${MYSQL_USERNAME}\"..."
      mysql --defaults-file=/etc/mysql/debian.cnf \
        -e "GRANT ALL PRIVILEGES on \`${MYSQL_DATABASE}\`.* TO '${MYSQL_USERNAME}' IDENTIFIED BY '${MYSQL_PASSWORD}';"
    fi
  fi

  /usr/bin/mysqladmin --defaults-file=/etc/mysql/debian.cnf shutdown
fi


services/20-mysql.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

MYSQL_DATABASE=${MYSQL_DATABASE-temp}
MYSQL_USERNAME=${MYSQL_USERNAME-mysql}
MYSQL_PASSWORD=${MYSQL_PASSWORD-mysql}

MYSQL_CHARSET=${MYSQL_CHARSET:-utf8}
MYSQL_COLLATION=${MYSQL_COLLATION:-utf8_unicode_ci}

MYSQL_USER=mysql
MYSQL_DATA_DIR=${MYSQL_DATA_DIR:-/var/lib/mysql}
MYSQL_LOG_DIR=/var/log/mysql/
MYSQL_RUN_DIR=/var/run/mysqld/

mkdir -p ${MYSQL_DATA_DIR}
chmod -R 0700 ${MYSQL_DATA_DIR}
chown -R ${MYSQL_USER}:${MYSQL_USER} ${MYSQL_DATA_DIR}

mkdir -p ${MYSQL_RUN_DIR}
chmod -R 0755 ${MYSQL_RUN_DIR}
chown -R ${MYSQL_USER}:root ${MYSQL_RUN_DIR}

mkdir -p ${MYSQL_LOG_DIR}
chmod -R 0755 ${MYSQL_LOG_DIR}
chown -R ${MYSQL_USER}:root ${MYSQL_LOG_DIR}

DISABLE_MYSQL=${DISABLE_MYSQL:-0}

if [ ! "${DISABLE_MYSQL}" -eq 0 ]; then
  touch /etc/service/mysql/down
else
  rm -f /etc/service/mysql/down
fi

if [ -f "/etc/mysql/conf.d/mysqld.cnf" ]; then
  echo "mysqld.cnf already exists"
else
  echo "[mysqld]" > /etc/mysql/conf.d/mysqld.cnf
  echo "datadir=${MYSQL_DATA_DIR}" >> /etc/mysql/conf.d/mysqld.cnf
fi

sed -i -e '/bind-address/d' /etc/mysql/mysql.conf.d/mysqld.cnf
if [ -f "/etc/mysql/conf.d/mysqld-skip-name-resolv.cnf" ]; then
  echo "mysqld-skip-name-resolv.cnf already exists"
else 
  cp /config/etc/mysql/conf.d/mysqld-skip-name-resolv.cnf /etc/mysql/conf.d/mysqld-skip-name-resolv.cnf
fi

if [ -f "/etc/mysql/conf.d/mysqld-bind-address.cnf" ]; then
  echo "mysqld-bind-address.cnf already exists"
else
  cp /config/etc/mysql/conf.d/mysqld-bind-address.cnf /etc/mysql/conf.d/mysqld-bind-address.cnf
fi

# Blank password for debian-sys-maint user
sed 's/password = .*/password =  /g' -i /etc/mysql/debian.cnf

if [ ! -d ${MYSQL_DATA_DIR}/mysql ]; then
  echo "Initializing database ..."
  mysqld --initialize-insecure --user=${MYSQL_USER} 
  echo "Starting MySQL Server ..."
  /usr/bin/mysqld_safe >/dev/null 2>&1 &

  timeout=30
  echo -n "Waiting for database server to accept connections"
  while ! /usr/bin/mysqladmin -u root status >/dev/null 2>&1
  do
    timeout=$(($timout - 1))
    if [ $timeout -eq 0 ]; then
      echo -e "\nCould not connect to database server. Aborting..."
      exit 1
    fi
    echo -n "."
    sleep 1
  done
  echo

  echo "Creating debian-sys-maint user ..."
  mysql -uroot -e "CREATE USER 'debian-sys-maint'@'localhost' IDENTIFIED BY '';"
  mysql -uroot -e "GRANT ALL PRIVILEGES on *.* TO 'debian-sys-maint'@'localhost' WITH GRANT OPTION;"

  /usr/bin/mysqladmin --defaults-file=/etc/mysql/debian.cnf shutdown
fi

if [ -n "${MYSQL_DATABASE}" -o -n "${MYSQL_USERNAME}" ]; then
  echo "Starting MySQL Server ..."
  /usr/bin/mysqld_safe >/dev/null 2>&1 &

  timeout=30
  echo -n "Waiting for database server to accept connections"
  while ! /usr/bin/mysqladmin -u root status >/dev/null 2>&1
  do
    timeout=$(($timout - 1))
    if [ $timeout -eq 0 ]; then
      echo -e "\nCould not connect to database server. Aborting..."
      exit 1
    fi
  done
  echo

  if [ -n "${MYSQL_DATABASE}" ]; then
    echo "Creating database ..."
    mysql --defaults-file=/etc/mysql/debian.cnf \
      -e "CREATE DATABASE IF NOT EXISTS \`${MYSQL_DATABASE}\` DEFAULT CHARACTER SET \`${MYSQL_CHARSET}\` COLLATE \`${MYSQL_COLLATION}\`;"

    if [ -n "${MYSQL_USERNAME}" ]; then
      echo "Granting access to database \"${MYSQL_DATABASE}\" for user \"${MYSQL_USERNAME}\"..."
      mysql --defaults-file=/etc/mysql/debian.cnf \
        -e "CREATE USER '${MYSQL_USERNAME}' IDENTIFIED BY '${MYSQL_PASSWORD}';"
      mysql --defaults-file=/etc/mysql/debian.cnf \
        -e "GRANT ALL PRIVILEGES on \`${MYSQL_DATABASE}\`.* TO '${MYSQL_USERNAME}';"
    fi
  fi

  /usr/bin/mysqladmin --defaults-file=/etc/mysql/debian.cnf shutdown
fi
